### PR TITLE
Add pause on exception to debugger

### PR DESCRIPTION
**Pull request by goanpeca**
## References

https://github.com/jupyterlab/jupyterlab/issues/9544

## User-facing changes

jlab

IPykernel 6.0rc1

image

### CODE REVIEW

**Comment by goanpeca** [comment]
Hi @JohanMabille I am trying to implement the "pause on exception functionality"

From the initialize method I get:

Screen Shot 2021-05-26 at 12 56 21

Then I added a method triggered by a pause button:

```ts
    header.toolbar.addItem(
      'pauseOnExceptions',
      new ToolbarButton({
        icon: pauseOnExceptionsIcon,
        onClick: async (): Promise<void> => {
          return service.pauseOnExceptions();
        },
        tooltip: trans.__('Pause on uncaught exceptions')
      })
    );
```

```ts
  async pauseOnExceptions(): Promise<void> {
    if (!this.session?.isStarted) {
      return
    }

    console.log("Pause on exceptions");
    const reply = await this.session.sendRequest(
      'setExceptionBreakpoints',
      {
        filters: ["raise", "uncaught"],
        exceptionOptions: [{ breakMode: "always"}, { breakMode: "always"}],
      });
    console.log(reply);
  }
```

The response has no `body`so I am not sure if things are working as expected on `xeus-python`

It says that a stopped event should be triggered but after enabling this and running some code that raises an exception, I am getting no event 

Screen Shot 2021-05-26 at 12 58 26

I appreciate it a lot if you could give me some pointers :). Thanks!

File: packages/debugger/src/service.ts [hunk 1]
```diff
@@ -468,7 +470,22 @@ export class DebuggerService implements IDebugger, IDisposable {
...
+  async pauseOnExceptions(): Promise<void> {
+    if (!this.session?.isStarted) {
+      return
+    }
+
+    console.log("Pause on exceptions");
+    const reply = await this.session.sendRequest(
+      'setExceptionBreakpoints',
+      {
+        filters: ["raise", "uncaught"],
```
Comment (JohanMabille) [inline packages/debugger/src/service.ts:485]: I think you have a typo here, it should be `raised` instead of `raise`.

Comment (goanpeca) [reply packages/debugger/src/service.ts:485]: Hi @JohanMabille, indeed I missed this, however still no event is triggered when an exception happens :-|

Comment (JohanMabille) [reply packages/debugger/src/service.ts]: I'm gonna investigate this on the backend side today.

Comment (goanpeca) [reply packages/debugger/src/service.ts]: Thanks! 🎊

**Comment by goanpeca** [comment]
That is great @JohanMabille 

Thanks a lot for the feedback. Will give it a try.

I got the impression path was optional from reading the docs 🤔

**Comment by JohanMabille** [comment]
> I got the impression path was optional from reading the docs thinking

Indeed, the field is optional in the protocol (and then in the generic use case). But it seems to be mandatory for debugpy.

**Comment by goanpeca** [comment]
>I've released xeus-python 0.12.5. The debug_info_response has a new field called exceptionPaths which is the list of string to use in the setExceptionBreakpoints request.

Thanks @JohanMabille

**Comment by JohanMabille** [comment]
My bad, I totally forgot the PyPI release! Fixing it now.

**Comment by JohanMabille** [comment]
Wheels are available on Linux and OSX, Windows build is on going.

**Comment by goanpeca** [comment]
Thanks @JohanMabille 

I have some questions :)

- Should ipykernel be updated too to return the additional information on the debug response?
- Since the ExceptionBreakpointFilters information that the server returns is going to be specific for, I am currently making some assumptions on using "raised" and "uncaught"
Screen Shot 2021-07-07 at 4 57 07 AM

Also I am not sure I understand why the default value for the `uncaught` filter is True, and for `raised` is False.

**Comment by JohanMabille** [comment]
> Should ipykernel be updated too to return the additional information on the debug response?

Yes, I will update it once we have a prototype that works.

> Also I am not sure I understand why the default value for the uncaught filter is True, and for raised is False.

I think it's the most common cas when debugging: you don't want to interrupt the code execution each time an exception is raised, but you want to break if the exception is not caught instead of having your code execution terminating.

> Would it be safe to assume for an initial pass that uncaught and raised are expected values in order ot enable disable this capability?

I think `raised` and `uncaught` are quite common values, however you should not rely on them to enable or disable the feature. The `Initialize` response contains a `Capabilities` body which contains an `exceptionBreakpointFilters` field when this feature is supported. I think we should rely on this instead.

**Comment by echarles** [comment]
@JohanMabille I read your last comment on https://github.com/jupyterlab/jupyterlab/pull/10298#issuecomment-876564714 and wonder what is missing to move this forward? I am not sure what the next action is and who should take it?

Otherwise, I had a look at the diff and I can understand what I read. It looks also that it is working with ipykernel, so a quick-win would be to rebase, remove the `console.log` and merge cc/ @andrewfulton9

**Comment by JohanMabille** [comment]
> @JohanMabille I read your last comment on [#10298 (comment)](https://github.com/jupyterlab/jupyterlab/pull/10298#issuecomment-876564714) and wonder what is missing to move this forward? I am not sure what the next action is and who should take it?

We only the PR I just opened in ipykernel to be merged, and then we should be able to move this forward.

**Comment by echarles** [comment]
@andrewfulton9 I have tested this branch. On my env, I have the following

1. The toggling of the "Pause on exeception" icon is not visible (as a user, I don't know if I will pause on not).
2. When pausing on exception, the horizontal orange bar is not shown anymore (it was shown in the initial gif, in the first comment of this issue).

Untitled

**Comment by echarles** [comment]
> The toggling of the "Pause on exeception" icon is not visible (as a user, I don't know if I will pause on not).

This is fixed via https://github.com/jupyterlab/jupyterlab/issues/11656

> When pausing on exception, the horizontal orange bar is not shown anymore (it was shown in the initial gif, in the first comment of this issue).

We still need to address that last one.

**Comment by andrewfulton9** [comment]
@echarles, I just pushed up some edits that add the horizontal orange bar back in. With that, this PR should be ready to go. Let me know what you think. It looks there is some issue with github actions though causing it to fail to reach yarn though.

**Comment by echarles** [comment]
horizontal bar is now shown 👍 I have manually relaunched a few failing CI jobs.

Screenshot 2021-12-16 at 23 17 14

**Comment by echarles** [comment]
@andrewfulton9 2 issues in CI

- Error: test/debugger.spec.ts(92,7): error TS2741: Property 'pause' is missing in type '{ registry: CommandRegistry; continue: string; terminate: string; next: string; stepIn: string; stepOut: string; evaluate: string; }' but required in type 'ICommands'.
-  echo 'Please run `jlpm run eslint` locally and push changes' Please run `jlpm run eslint` locally and push changes

I have also looked at the button, and it does not show the state although the other buttons are shown. Have you rebased to pick the latests CSS changes from https://github.com/jupyterlab/jupyterlab/pull/11688 or is there something to do for that pause on exception button?

Screenshot 2021-12-17 at 07 36 38

**Comment by andrewfulton9** [comment]
@echarles, I think these errors may have been introduced when merging into the branch used for this PR. I think at this point it will be easier if I create a new PR referencing this one, that I can directly push to.

**Comment by echarles** [comment]
> @echarles, I think these errors may have been introduced when merging into the branch used for this PR. I think at this point it will be easier if I create a new PR referencing this one, that I can directly push to.

Sure, I let you open a new PR and close this one. Thx again @andrewfulton9

**Comment by andrewfulton9** [comment]
@echarles, new PR is #11752. This PR can be closed now I think. Thanks!

**Comment by echarles** [comment]
Closing in favor of https://github.com/jupyterlab/jupyterlab/pull/11752

**Pull request closed without merging**